design/004-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,255 @@
+external sources. The primary concern to address is how to do justice to
+external data with respect to provenance tracking; Renku should preserve the
+information about the original source of the data and provide mechanisms to
+retrieve this data when necessary.
+
+### User stories
+
+#### 1. Sharing data for ad-hoc analysis
+
+This example uses S3 but it could be nearly identical with e.g. an institutional
+NFS server or something like openBIS.
+
+Bob is a scientist working in a large international collaboration that relies on
+sharing raw data that is very valuable, large (TB-range), and immutable. Bob
+starts looking at the data on his laptop. The data is in an S3 bucket that his
+lab uses. He uses s3fuse to mount the S3 bucket with the data to his laptop.
+
+He then decides that he needs to give others access to his work-in-progress so
+they can decide on future direction for the analysis. He wants people to have
+access to the same compute runtime so there are no conflicts with libraries that
+are used. He creates a renku project on his laptop, updates the list of
+dependencies and creates a dataset. He adds the data from S3 to his Renku
+dataset; Renku queries him for credentials to be able to access the protected S3
+bucket and they are stored in his local user configuration so he doesn't have to
+enter them again. Bob uses the Renku CLI to mount the data to a directory in
+his project so it can be used seamlessly like any other data or file that he
+might have. He uses the data in his project to prepare some simple sanity checks
+to share with his collaborators. He annotates the dataset with metadata so they can
+more easily find it later if needed for another project.
+
+Bob pushes the project to renkulab.io and shares it with his peers. The images are
+built automatically and include the software packages needed for the analysis that
+Bob created. He and his colleagues can launch interactive sessions for the project
+and Renku automatically prompts them to mount the data from S3 that is used
+in the dataset. Once the session starts, the S3 bucket is mounted automatically and
+it is available in the same location in the project as it was before on Bob's
+personal laptop.
+
+
+#### 2. Tracking the use of external data
+
+Sally is Bob's collaborator who continues to build on Bob's initial work by
+building a more robust workflow. She uses standard Renku commands to record the
+workflows. Renku uses the same mechanisms to attach the data as described above
+when workflows are being re-executed or managed. She can also request to check
+whether the results are still in sync with the data using `renku status`. In
+this case, Renku checks the object hashes that were recorded previously against
+the data currently found in the remote location (this might take a while over
+the network - some data sources might provide pre-computed hashes). If the
+hashes differ, Renku offers to update the results.
+
```

Comment:
```suggestion

#### 3. Data storage for research project

Willow is interested in a large, protected dataset that is offered for free to researchers by an external organization. She has to accept a data access agreement to access the data, at which points she receives credentials to the bucket. To explore the dataset, she creates a Renku project and mounts the S3 bucket using the credentials.

After some exploration of the dataset, Willow decides to start a research project with the dataset. She makes a copy of the dataset in her own private S3 bucket for two reasons: (1) to make sure she maintains access to the data independent of the external organization, and (2) because she prefers to keep the project’s raw, intermediate, and final data products in the same place. She creates a Renku project for her work and mounts the bucket with her credentials. Throughout the research project, she reads the raw data from the mounted directory and saves intermediate data files and figures in separate folders in the same S3 bucket.

```